cli/fossilize_list.cpp

```diff
@@ -26,16 +26,278 @@
 #include "layer/utils.hpp"
 #include <memory>
 #include <vector>
+#include <unordered_set>
+#include <map>
 
 using namespace Fossilize;
 using namespace std;
 
+#define PRINT_SAVED_BUFFER(SavedBufferName, VkObjectName, Tag)\
+for (auto It = SavedBufferName.begin(); It != SavedBufferName.end(); ++It)\
+{\
+	printf(VkObjectName "(%d):%016" PRIx64 ", ", Tag, *It);\
+}\
+
+static const ResourceTag playback_order[] = {
+	RESOURCE_SAMPLER,
+	RESOURCE_DESCRIPTOR_SET_LAYOUT,
+	RESOURCE_PIPELINE_LAYOUT,
+	RESOURCE_SHADER_MODULE,
+	RESOURCE_RENDER_PASS,
+	RESOURCE_GRAPHICS_PIPELINE,
+	RESOURCE_COMPUTE_PIPELINE,
+	RESOURCE_RAYTRACING_PIPELINE
+};
+
 static void print_help()
 {
 	LOGI("Usage: fossilize-list\n"
 	     "\t<database path>\n"
 	     "\t[--tag index]\n"
-	     "\t[--size]\n");
+	     "\t[--size]\n"
+		 "\t[--connectivity]\n");
+}
+
+template <typename T>
+static inline T fake_handle(uint64_t v)
+{
+	return (T)v;
+}
+
+template <typename T>
+static inline const T* find_pnext(VkStructureType type, const void* pNext)
+{
+	while (pNext != nullptr)
+	{
+		auto* sin = static_cast<const VkBaseInStructure*>(pNext);
+		if (sin->sType == type)
+			return static_cast<const T*>(pNext);
+
+		pNext = sin->pNext;
+	}
+
+	return nullptr;
+}
+
+struct ListReplayer : StateCreatorInterface
+{
+	struct SavedHashes
+	{
+		unordered_set<Hash> saved_samplers;
+		unordered_set<Hash> saved_descriptor_sets;
+		unordered_set<Hash> saved_pipeline_layouts;
+		unordered_set<Hash> saved_shader_modules;
+		unordered_set<Hash> saved_render_passes;
+		unordered_set<Hash> saved_graphics_pipelines;
+		unordered_set<Hash> saved_raytracing_pipelines;
+	};
+
+	map<Hash, SavedHashes> saved_hashes_map;
```

Comment:
Why is this an ordered map?

Reply:
Also, you need different hash maps per resource type. Hashes from difference resource types are allowed to collide.